### PR TITLE
fix(npm-resolver): unhandled rejection in fetch

### PR DESCRIPTION
**Pull request by amitgilad3**
When fetch command fails for any reason catch it and return FetchError to avoid Unhandled rejection

ref #3261

### CODE REVIEW

File: packages/npm-resolver/src/fetch.ts [hunk 1]
```diff
@@ -50,11 +50,21 @@ export default async function fromRegistry (
...
-      }) as RegistryResponse
+      let response: RegistryResponse
+      try {
+        response = await fetch(uri, {
+          authHeaderValue,
+          retry: fetchOpts.retry,
+          timeout: fetchOpts.timeout,
+        }) as RegistryResponse
+      } catch (error) {
+        console.log(error)
```
Comment (zkochan) [inline packages/npm-resolver/src/fetch.ts]: this is not needed

File: packages/error/src/index.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export default class PnpmError extends Error {
   }
 }
 
-export interface FetchErrorResponse { status: number, statusText: string }
+export interface FetchErrorResponse { status: number | string, statusText: string }
```
Comment (zkochan) [inline packages/error/src/index.ts]: This error object was created for errors that have an http status respond, so I think it is better no to change it.

File: packages/npm-resolver/src/fetch.ts [hunk 3]
```diff
@@ -50,11 +50,21 @@ export default async function fromRegistry (
...
+        reject(new FetchError({
+          url: uri,
+          authHeaderValue: authHeaderValue,
+        }, { status: error.code, statusText: error.message }))
```
Comment (zkochan) [inline packages/npm-resolver/src/fetch.ts]: Probably best to use the generic error object

```ts
reject(new PnpmError('META_FETCH_FAIL', `GET ${uri}: ${err.message}`, { attempts: attempt }))
```

**Review by zkochan — CHANGES_REQUESTED** [review]

File: packages/npm-resolver/src/fetch.ts (hunk 1 above)
Comment (amitgilad3) [reply packages/npm-resolver/src/fetch.ts]: my mistake 👍

File: .changeset/little-dragons-wave.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@pnpm/error": patch
```
Comment (zkochan) [inline .changeset/little-dragons-wave.md]: This line should be removed.

File: packages/error/src/index.ts (hunk 2 above)
Comment (amitgilad3) [reply packages/error/src/index.ts]: thanks for the explanation , makes perfect sense

**Review by zkochan — APPROVED** [review]

**Comment by amitgilad3** [comment]
sorry for the inconvenience , this is my first pr , will pay more attention to errors in the future.
i fixed the failing tests but locally on branch main not all test pass (without my fixes),  is that the current state ? @zkochan

**Comment by zkochan** [comment]
Tests seem to pass on main. What is your OS?

**Comment by amitgilad3** [comment]
macos

**Pull request merged**